### ISSUE TITLE
Promtail won't stop when loki isn't running

### ISSUE THREAD

**Issue by rightaway** [grafana/loki#6533]
**Describe the bug**
When loki isn't running and you stop promtail it doesn't stop itself.

**To Reproduce**
1. start loki and promtail (version 2.5.0)
2. stop loki
3. stop promtail

**Expected behavior**
promtail should stop immediately

**Screenshots, Promtail config, or terminal output**
Promtail doesn't stop. If running `promtail-linux-amd64` on the command line you have to enter ctrl C twice. Or if through systemd it takes 90 seconds for the stop request to time out and kill it. That's a problem when restarting the server because the restart is blocked for that time.

When you request to stop promtail this appears in the log

```
caller=signals.go:55 msg="=== received SIGINT/SIGTERM ===\n*** exiting"
```

But promtail logs show this message every few seconds before and after that

```
caller=client.go:349 component=client host=127.0.0.1:3100 msg="error sending batch, will retry" status=-1 error="Post \"http://127.0.0.1:3100/loki/api/v1/push\": dial tcp 127.0.0.1:3100: connect: connection refused"
```

If you stop promtail when loki is running then it stops immediately. But promtail should stop immediately when requested even if loki is not running.

**Comment by rightaway** [comment]
> Promtail is actually not supposed to stop immediately to ensure no data is lost

@jeschkies Why would anything be lost? The positions file can help make sure it doesn't lose anything.

> Maybe we should document this better

It's not a documentation problem it's a behavior problem. If an administrator has requested the process to stop it's because they need it to stop right away. That's how every process works.

Getting in the way of that causes problems. For example if you reboot a server, every process behaves normally and the server should restart within seconds, but because you're using promtail you're locked out of the server for a while because promtail hasn't shut down and it can't reboot yet.

Nobody needs surprises in production and nobody needs promtail thinking it knows more than the administrator who needs a process to end for whatever reason.

> You could reduce the number of retries.

Retries are important during normal operation and there's no point to disable them.

When promtail gets a shutdown command it needs to be in shutdown mode and follow different rules.

**Comment by rightaway** [comment]
Retrying during normal operation is good but not during shutdown mode. It could log that it wasn't able to reach loki and shutdown immediately.

Or add a setting to disable retries only in shutdown mode so you don't have to disable it all the time.

**Comment by jeschkies** [comment]
We should make it configurable then. So the option would be no-retries on `SIGTERM`.

**Comment by frittentheke** [comment]
> > Why would anything be lost?
> 
> Depending on the target you use some data is actually in-memory and Promtail will retry until it gives up. The position file will not help there. The Docker driver has similar behaviour (#2361). You could send a SIGKILL instead. That will shut down Promtail immediately.

Yes, I believe the issue here is that the file positions are advanced to a point that relates to still in-flight data. And if that data, even with retries, cannot be pushed to Loki that data is lost when shutting down Promtail.

I am wondering why not advance the positions only after successful pushes to Loki instead of "just" reading that data into Promtails pipeline?

**Comment by rightaway** [comment]
@vitaliyf It's a workaround but I think the default behavior is better to change for the reasons discussed.

**Comment by damluk** [comment]
> I am wondering why not advance the positions only after successful pushes to Loki instead of "just" reading that data into Promtails pipeline?

While this might be a solution for logfile scraping, a syslog receiver, even TCP-based, is in trouble because data that the TCP/IP stack had already ACKed, will be lost if it is only kept in memory. Promtail needs a persistent queue.

**Comment by ku1ik** [comment]
I don't care if I lose some logs at the time of promtail shutdown so I used  [TimeoutStopSec](https://www.freedesktop.org/software/systemd/man/systemd.service.html#TimeoutStopSec=) as a workaround in promtail.service unit:

```ini
[Service]
...
TimeoutStopSec=5
```

Works fine.

**Comment by cstyan** [comment]
Hello, thanks for your feature request.

As a general statement: we're currently reevaluating promtails position as a project within Grafana Labs. Internally we're actually using the [Agent](https://github.com/grafana/agent) for both metrics and logs collection at this point.

While we haven't made a formal decision yet, we expect in the near future that all new feature work will be done in the Agent's log collection pipelines rather than in Promtail.

To add some more detail for this case specifically, I also maintain Prometheus remote write. While both Prometheus' WAL and on disk log files can act like semi-permanent buffers, as others have mentioned we don't have anything like positions files for logs data coming from somewhere other than log files, and we don't have something like positions files for prometheus' WAL when used by remote write. Promtails WAL is also very basic and would need aditional work to support that use case. Without these changes, I don't think we should default to immediate shutdown or remove the current behaviour. Capturing a signal like ctrl+c and attempting to do a graceful shutdown before doing a hard exit is a common pattern.

**Comment by damluk** [comment]
> As a general statement: we're currently reevaluating promtails position as a project within Grafana Labs. Internally we're actually using the [Agent](https://github.com/grafana/agent) for both metrics and logs collection at this point.
> 
> While we haven't made a formal decision yet, we expect in the near future that all new feature work will be done in the Agent's log collection pipelines rather than in Promtail.

@cstyan, "Grafana Agent has been deprecated and is now in Long-Term Support mode". Can you please update your previous statement regarding promtail?

**Comment by cstyan** [comment]
@damluk the same readme you're quoting specifies `We recommend migrating to the new [Grafana Alloy](https://github.com/grafana/alloy) collector, which is built on the foundation of Grafana Agent Flow.`, which in reality is just the agent rebranded as `Alloy`. The rest of my statement remains the same, Promtail is now officially in a long term maintenance state and will not recieve new features, we recommend using Alloy.

**Comment by thequailman** [comment]
For folks coming to this issue and being steered towards Alloy: don't, it's a train wreck.  It's very bloated, and the config syntax is more like HCL for some asinine reason and extremely difficult to understand compared to something like node_exporter and promtail.  Idk what Grafana is thinking here, and it's unfortunate that promtail may be being abandoned.   Depending on your needs, I found that Netdata is a CNCF project now and can handle most of the "Grafana stack" stuff better: https://github.com/netdata/netdata.

**Comment by JStickler** [comment]
Promtail is end of life (EOL) as of  March 2, 2026.  No future support or updates will be provided. All future feature development will occur in [Grafana Alloy](https://github.com/grafana/alloy).
